### PR TITLE
Encryptionboxes

### PR DESCRIPTION
**Pull request by Amphyby**

### CODE REVIEW

File: src/debot/term_encryption_box.rs [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+            let pair = load_keypair(&value).map_err(|e| e.to_string())?;
+            key = format!("{:064}", pair.secret);
+        }
+
+        let registered_box = match params.box_type {
+            EncryptionBoxType::SecretNaCl => {
+                register_encryption_box(
+                    params.context.clone(),
+                    NaClSecretBox {
+                        key: hex::encode(&key),
```
[the remark this answers is not in the document]
Comment (Amphyby) [reply src/debot/term_encryption_box.rs]: removed

File: src/debot/term_encryption_box.rs [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+                .map_err(|e| e.to_string())?
+                .handle
+            }
+            EncryptionBoxType::NaCl => {
+                let padded_pubkey = format!("{:064}", params.their_pubkey);
+                register_encryption_box(
+                    params.context.clone(),
+                    NaClBox {
+                        their_pubkey: hex::encode(&padded_pubkey),
+                        secret: hex::encode(&key),
```
[the remark this answers is not in the document]
Comment (Amphyby) [reply src/debot/term_encryption_box.rs]: removed

File: src/debot/term_encryption_box.rs [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+                .await
+                .map_err(|e| e.to_string())?
+                .handle
+            }
+            EncryptionBoxType::ChaCha20 => {
+                register_encryption_box(
+                    params.context.clone(),
+                    ChaChaBox {
+                        key: base64::encode(&key),
+                        nonce: hex::encode(&params.nonce),
```
[the remark this answers is not in the document]
Comment (Amphyby) [reply src/debot/term_encryption_box.rs]: removed

File: src/debot/term_encryption_box.rs [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+                )
+                .await
+                .map_err(|e| e.to_string())?
+                .handle
+            }
+            EncryptionBoxType::ChaCha20 => {
+                register_encryption_box(
+                    params.context.clone(),
+                    ChaChaBox {
+                        key: base64::encode(&key),
```
[the remark this answers is not in the document]
Comment (Amphyby) [reply src/debot/term_encryption_box.rs]: removed

[the remark this answers is not in the document]
Comment (Amphyby) [reply src/debot/term_encryption_box.rs]: removed

File: src/debot/term_encryption_box.rs [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+                .await
+                .map_err(|e| e.to_string())?
+                .handle
+            }
+            EncryptionBoxType::NaCl => {
+                let padded_pubkey = format!("{:064}", params.their_pubkey);
+                register_encryption_box(
+                    params.context.clone(),
+                    NaClBox {
+                        their_pubkey: hex::encode(&padded_pubkey),
```
[the remark this answers is not in the document]
Comment (Amphyby) [reply src/debot/term_encryption_box.rs]: removed

File: src/debot/interfaces/encryption_box_input.rs [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        let prompt = decode_prompt(args)?;
+        let nonce = decode_nonce(args)?;
+        println!("{}", prompt);
+        let encryption_box = TerminalEncryptionBox::new(ParamsOfTerminalEncryptionBox {
+            context: self.client.clone(),
+            box_type: EncryptionBoxType::SecretNaCl,
+            their_pubkey: String::new(),
+            nonce: nonce,
+        })
+        .await?;
```
[the remark this answers is not in the document]
Comment (Keshoid) [reply src/debot/interfaces/encryption_box_input.rs:119]: See comment for TerminalEncryptionBox::new()

File: src/debot/interfaces/encryption_box_input.rs [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        let handle = encryption_box.handle();
+        self.handles.write().await.push(encryption_box);
+        Ok((answer_id, json!({ "handle": handle.0})))
+    }
+    async fn remove_handle(&self, args: &Value) -> InterfaceResult {
+        let answer_id = decode_answer_id(args)?;
+        let handle = decode_num_arg::<u32>(args, "handle")?;
+        let initial_size = self.handles.write().await.len();
```
Comment (Keshoid) [inline src/debot/interfaces/encryption_box_input.rs:143]: Get Write Lock only once. Then do all operations like .len(), .retain(), .len():
let handles = self.handles.write().await;
handles.len();
handles.retain(...);
handles.len();

File: src/debot/term_encryption_box.rs [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+        .decrypted)
+    }
+}
+
+#[async_trait::async_trait]
+impl ton_client::crypto::EncryptionBox for ChaChaBox {
+    async fn get_info(&self) -> ClientResult<EncryptionBoxInfo> {
+        Ok(EncryptionBoxInfo {
+            hdpath: Some(String::from(HD_PATH)),
+            algorithm: Some("ChaCha20".to_string()),
```
Comment (Keshoid) [inline src/debot/term_encryption_box.rs:96]: I prefer `format!("ChaCha20")` instead of "..".to_string().

File: src/debot/term_encryption_box.rs [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+                    params.context.clone(),
+                    NaClSecretBox {
+                        key: key,
+                        nonce: params.nonce,
+                        client: params.context.clone(),
+                    },
+                )
+                .await
+                .map(|r| r.handle)
+                .unwrap_or(EncryptionBoxHandle(0))
```
Comment (Keshoid) [inline src/debot/term_encryption_box.rs:197]: Move this 2 lines to the caller function and write here: `.await?;`

File: src/debot/term_encryption_box.rs [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+                    NaClBox {
+                        their_pubkey: params.their_pubkey,
+                        secret: key,
+                        nonce: params.nonce,
+                        client: params.context.clone(),
+                    },
+                )
+                .await
+                .map(|r| r.handle)
+                .unwrap_or(EncryptionBoxHandle(0))
```
Comment (Keshoid) [inline src/debot/term_encryption_box.rs:211]: Move this 2 lines to the caller function and write here: .await?;

File: src/debot/term_encryption_box.rs [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+                    params.context.clone(),
+                    ChaChaBox {
+                        key: key,
+                        nonce: params.nonce,
+                        client: params.context.clone(),
+                    },
+                )
+                .await
+                .map(|r| r.handle)
+                .unwrap_or(EncryptionBoxHandle(0))
```
Comment (Keshoid) [inline src/debot/term_encryption_box.rs:224]: Move this 2 lines to the caller function and write here: .await?;

File: src/debot/interfaces/encryption_box_input.rs [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+        self.handles
+            .write()
+            .await
+            .retain(|value| (*value).handle().0 != handle);
+        let removed: bool = initial_size != self.handles.write().await.len();
```
Comment (Keshoid) [inline src/debot/interfaces/encryption_box_input.rs:148]: You didn't close the handle. Just dropped it. To really close the handle you should call `remove_encryption_box`.
Check https://github.com/tonlabs/tonos-cli/blob/encryptionboxes/src/debot/term_signing_box.rs#L51 how to automatically close handle when `TerminalEncryptionBox` will be destroyed.

**Review by Keshoid — CHANGES_REQUESTED** [review]

File: src/debot/interfaces/dinterface.rs [hunk 13]
```diff
@@ -96,7 +100,11 @@ pub fn decode_string_arg(args: &Value, name: &str) -> Result<String, String> {
     let bytes = hex::decode(&decode_arg(args, name)?).map_err(|e| format!("{}", e))?;
     std::str::from_utf8(&bytes)
         .map_err(|e| format!("{}", e))
-        .map(|x| x.to_string())
+        .map(|x|  x.to_string())
```
[the remark this answers is not in the document]
Comment (Amphyby) [reply src/debot/interfaces/dinterface.rs:103]: done

File: src/debot/interfaces/encryption_box_input.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Amphyby) [reply src/debot/interfaces/encryption_box_input.rs:119]: done

File: src/debot/interfaces/encryption_box_input.rs (hunk 7 above)
Comment (Amphyby) [reply src/debot/interfaces/encryption_box_input.rs:143]: done

File: src/debot/term_encryption_box.rs (hunk 8 above)
Comment (Amphyby) [reply src/debot/term_encryption_box.rs:96]: done

File: src/debot/term_encryption_box.rs (hunk 10 above)
Comment (Amphyby) [reply src/debot/term_encryption_box.rs:211]: done

File: src/debot/term_encryption_box.rs (hunk 9 above)
Comment (Amphyby) [reply src/debot/term_encryption_box.rs:197]: done

File: src/debot/term_encryption_box.rs (hunk 11 above)
Comment (Amphyby) [reply src/debot/term_encryption_box.rs:224]: done

File: src/debot/interfaces/encryption_box_input.rs (hunk 12 above)
Comment (Amphyby) [reply src/debot/interfaces/encryption_box_input.rs]: done

**Review by Keshoid — APPROVED** [review]

**Pull request merged**